File: seed/agency/deskhr/lib/rbac.ts
// rbac.ts — role-based access control.
//
// Direct port of dashboard/rbac.py. The same PERMISSIONS matrix gates both the
// API routes (server-side enforcement) and the dashboard widgets (client-side
// hiding) — defense in depth, exactly as the Python reference describes.

import type { RoleName } from '@/lib/data';

export const Permission = {
  VIEW_OWN_SCHEDULE: 'view_own_schedule',
  VIEW_TEAM_ANALYTICS: 'view_team_analytics',
  VIEW_COMPANY_ANALYTICS: 'view_company_analytics',
  EDIT_SCHEDULE: 'edit_schedule',
  MANAGE_EMPLOYEES: 'manage_employees',
} as const;

export type PermissionName = (typeof Permission)[keyof typeof Permission];

// Role -> permissions. Higher roles are supersets, kept explicit for clarity.
export const PERMISSIONS: Record<RoleName, PermissionName[]> = {
  employee: [Permission.VIEW_OWN_SCHEDULE],
  manager: [
    Permission.VIEW_OWN_SCHEDULE,
    Permission.VIEW_TEAM_ANALYTICS,
    Permission.EDIT_SCHEDULE,
  ],
  hr_admin: [
    Permission.VIEW_OWN_SCHEDULE,
    Permission.VIEW_TEAM_ANALYTICS,
    Permission.VIEW_COMPANY_ANALYTICS,
    Permission.EDIT_SCHEDULE,
    Permission.MANAGE_EMPLOYEES,
  ],
};

export function hasPermission(role: RoleName, permission: PermissionName): boolean {
  return PERMISSIONS[role]?.includes(permission) ?? false;
}

export const ROLES: { value: RoleName; label: string }[] = [
  { value: 'employee', label: 'Employee' },
  { value: 'manager', label: 'Manager' },
  { value: 'hr_admin', label: 'HR Admin' },
];

export function isRole(value: string | null): value is RoleName {
  return value === 'employee' || value === 'manager' || value === 'hr_admin';
}

File: seed/agency/deskhr/dashboard/rbac.py
"""rbac.py — role-based access control.

A small permission matrix plus a view decorator. The same PERMISSIONS map is
exported to the frontend so `web/Dashboard.tsx` can hide widgets the viewer
isn't allowed to see (defense-in-depth: the API still enforces it here).
"""

from __future__ import annotations

import functools

from django.http import JsonResponse

from models.employee import Role


class Permission:
    VIEW_OWN_SCHEDULE = "view_own_schedule"
    VIEW_TEAM_ANALYTICS = "view_team_analytics"
    VIEW_COMPANY_ANALYTICS = "view_company_analytics"
    EDIT_SCHEDULE = "edit_schedule"
    MANAGE_EMPLOYEES = "manage_employees"


# Role -> set of permissions. Higher roles are supersets, kept explicit for clarity.
PERMISSIONS: dict[str, set[str]] = {
    Role.EMPLOYEE: {
        Permission.VIEW_OWN_SCHEDULE,
    },
    Role.MANAGER: {
        Permission.VIEW_OWN_SCHEDULE,
        Permission.VIEW_TEAM_ANALYTICS,
        Permission.EDIT_SCHEDULE,
    },
    Role.HR_ADMIN: {
        Permission.VIEW_OWN_SCHEDULE,
        Permission.VIEW_TEAM_ANALYTICS,
        Permission.VIEW_COMPANY_ANALYTICS,
        Permission.EDIT_SCHEDULE,
        Permission.MANAGE_EMPLOYEES,
    },
}


def has_permission(role: str, permission: str) -> bool:
    return permission in PERMISSIONS.get(role, set())


def require_permission(permission: str):
    """Decorator for Django views. Expects request.user.employee with a `.role`."""

    def decorator(view_func):
        @functools.wraps(view_func)
        def wrapper(request, *args, **kwargs):
            employee = getattr(request.user, "employee", None)
            if employee is None:
                return JsonResponse({"error": "unauthenticated"}, status=401)
            if not has_permission(employee.role, permission):
                return JsonResponse({"error": "forbidden", "need": permission}, status=403)
            return view_func(request, *args, **kwargs)

        return wrapper

    return decorator
